src/toil_lib/programs.py

```diff
@@ -88,29 +128,63 @@ def docker_call(tool,
 
     _log.debug("Calling docker with %s." % " ".join(base_docker_call + [tool] + parameters))
 
-    docker_call = base_docker_call + [tool] + parameters
+    call = base_docker_call + [tool] + parameters
 
-    try:
-        if outfile:
-            subprocess.check_call(docker_call, stdout=outfile)
+    if outfile:
+        subprocess.check_call(call, stdout=outfile)
+    else:
+        if check_output:
+            return subprocess.check_output(call)
         else:
-            if check_output:
-                return subprocess.check_output(docker_call)
-            else:
-                subprocess.check_call(docker_call)
+            subprocess.check_call(call)
     # Fix root ownership of output files
-    except:
-        # Panic avoids hiding the exception raised in the try block
-        with panic():
-            _fix_permissions(base_docker_call, tool, work_dir)
-    else:
-        _fix_permissions(base_docker_call, tool, work_dir)
+    _fix_permissions(base_docker_call, tool, work_dir)
 
     for filename in outputs.keys():
         if not os.path.isabs(filename):
             filename = os.path.join(work_dir, filename)
         assert(os.path.isfile(filename))
 
+docker_call.FORGO = 0
+docker_call.STOP = 1
+docker_call.RM = 2
+
+
+def _docker_kill(container_name, action):
+    """
+    Kills the specified container.
+
+    :param str container_name: The name of the container created by docker_call
+    :param int action: What action should be taken on the container?  See `defer=` in
+           :func:`docker_call`
+    """
+    running = _container_is_running(container_name)
+    if running is None:
+        # This means that the container doesn't exist.  We will see this if the container was run
+        # with --rm and has already exited before this call.
+        _log.info('The container with name "%s" appears to have already been removed.  Nothing to '
+                  'do.', container_name)
+        return
+    else:
+        if action in (None, docker_call.FORGO):
+            _log.info('The container with name %s continues to exist as we were asked to forgo a '
+                      'post-job action on it.', container_name)
+            return
+        else:
+            if running and action >= docker_call.STOP:
+                _log.info('Stopping container "%s".', container_name)
+                subprocess.check_call(['docker', 'stop', container_name])
+            if action >= docker_call.RM:
+                # If the container was run with --rm, then stop will most likely remove the
+                # container.  We first check if it is running then remove it.
+                running = _container_is_running(container_name)
+                if running is not None:
```

Comment:
Same here. If container is already gone, log that you think nothing needs to be done.